processor/queuedprocessor/queued_processor.go

```diff
@@ -16,76 +16,209 @@ package queuedprocessor
+	// TODO: Remove this after #1312 is merged and verify.
+	ctx, _ = tag.New(ctx, tag.Upsert(processor.TagProcessorNameKey, instanceName))
+	_, numPoints := pdatautil.MetricAndDataPointCount(md)
+	return &metricsQueueItem{
+		baseQueueItem: baseQueueItem{ctx: ctx, qt: time.Now()},
+		md:            md,
+		numPoints:     numPoints,
+	}
+}
+
+func (item *metricsQueueItem) onAccepted() {
+	obsreport.ProcessorMetricsDataAccepted(item.ctx, item.numPoints)
+}
+
+func (item *metricsQueueItem) onPartialError(consumererror.PartialError) {
+	// TODO: implement this.
+}
+
+func (item *metricsQueueItem) onRefused(logger *zap.Logger, err error) {
+	obsreport.ProcessorMetricsDataRefused(item.ctx, item.numPoints)
+
+	logger.Error("Failed to process batch, refused", zap.Int("#points", item.numPoints), zap.Error(err))
+}
+
+func (item *metricsQueueItem) onDropped(logger *zap.Logger, err error) {
+	stats.Record(item.ctx, processor.StatTraceBatchesDroppedCount.M(int64(1)))
+	obsreport.ProcessorMetricsDataDropped(item.ctx, item.numPoints)
+
+	logger.Error("Failed to process batch, discarding", zap.Int("#points", item.numPoints), zap.Error(err))
+}
+
+func (item *metricsQueueItem) export(sp *queuedProcessor) error {
+	return sp.metricNext.ConsumeMetrics(item.ctx, item.md)
+}
+
+func newQueuedTracesProcessor(
 	params component.ProcessorCreateParams,
-	sender consumer.TraceConsumer,
+	nextConsumer consumer.TraceConsumer,
 	cfg *Config,
-) *queuedSpanProcessor {
-	return &queuedSpanProcessor{
+) *queuedProcessor {
+	return &queuedProcessor{
 		name:                     cfg.Name(),
 		queue:                    queue.NewBoundedQueue(cfg.QueueSize, func(item interface{}) {}),
 		logger:                   params.Logger,
 		numWorkers:               cfg.NumWorkers,
-		sender:                   sender,
-		statsTags:                []tag.Mutator{tag.Insert(processor.TagProcessorNameKey, cfg.Name())},
+		traceNext:                nextConsumer,
+		metricNext:               nil,
+		retryOnProcessingFailure: cfg.RetryOnFailure,
+		backoffDelay:             cfg.BackoffDelay,
+		stopCh:                   make(chan struct{}),
+	}
+}
+
+func newQueuedMetricsProcessor(
+	params component.ProcessorCreateParams,
+	nextConsumer consumer.MetricsConsumer,
+	cfg *Config,
+) *queuedProcessor {
+	return &queuedProcessor{
+		name:                     cfg.Name(),
+		queue:                    queue.NewBoundedQueue(cfg.QueueSize, func(item interface{}) {}),
+		logger:                   params.Logger,
+		numWorkers:               cfg.NumWorkers,
+		traceNext:                nil,
+		metricNext:               nextConsumer,
 		retryOnProcessingFailure: cfg.RetryOnFailure,
 		backoffDelay:             cfg.BackoffDelay,
 		stopCh:                   make(chan struct{}),
 	}
 }
 
 // Start is invoked during service startup.
-func (sp *queuedSpanProcessor) Start(_ context.Context, _ component.Host) error {
+func (sp *queuedProcessor) Start(ctx context.Context, _ component.Host) error {
```

Comment:
nit: not from your change but perhaps a follow up to rename the receiver

Reply:
what name should I use?

Reply:
either `qp`or simply `q` it is fine for me, if you want to go further rename the type to `queuedRetry` (`qr`)